packages/hebao_v3/contracts/base/SmartWallet.sol

```diff
@@ -0,0 +1,546 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2017 Loopring Technology Limited.
+pragma solidity ^0.7.0;
+pragma experimental ABIEncoderV2;
+
+import "../lib/EIP712.sol";
+import "../lib/ERC20.sol";
+import "../lib/ERC1271.sol";
+import "../lib/ReentrancyGuard.sol";
+import "../thirdparty/ens/ENS.sol";
+
+import "./libwallet/ERC20Lib.sol";
+import "./libwallet/ERC1271Lib.sol";
+import "./libwallet/WalletData.sol";
+import "./libwallet/LockLib.sol";
+import "./libwallet/GuardianLib.sol";
+import "./libwallet/InheritanceLib.sol";
+import "./libwallet/MetaTxLib.sol";
+import "./libwallet/WhitelistLib.sol";
+import "./libwallet/QuotaLib.sol";
+import "./libwallet/RecoverLib.sol";
+import "./libwallet/UpgradeLib.sol";
+
+
+/// @title SmartWallet
+/// @dev Main smart wallet contract
+/// @author Brecht Devos - <brecht@loopring.org>
+contract SmartWallet is ERC1271
+{
+    using ERC20Lib          for Wallet;
+    using ERC1271Lib        for Wallet;
+    using LockLib           for Wallet;
+    using GuardianLib       for Wallet;
+    using InheritanceLib    for Wallet;
+    using MetaTxLib         for Wallet;
+    using WhitelistLib      for Wallet;
+    using QuotaLib          for Wallet;
+    using RecoverLib        for Wallet;
+    using RecoverLib        for Wallet;
+    using UpgradeLib        for Wallet;
+
+    bytes32     public immutable DOMAIN_SEPARATOR;
+    PriceOracle public immutable priceOracle;
+    address     public immutable walletFactory;
+
+    // WARNING: Do not delete wallet state data to make this implementation
+    // compatible with early versions.
+    //
+    //  ----- DATA LAYOUT BEGINS -----
+    // Always needs to be first
+    address internal masterCopy;
+
+    bool internal isImplementationContract;
+
+    Wallet public wallet;
+    //  ----- DATA LAYOUT ENDS -----
+
+    /// @dev We need to make sure the implemenation contract cannot be initialized
+    ///      and used to do delegate calls to arbitrary contracts.
+    modifier disableInImplementationContract
+    {
+        require(!isImplementationContract, "DISALLOWED_ON_IMPLEMENTATION_CONTRACT");
+        _;
+    }
+
+    modifier onlyFromWalletOrOwnerWhenUnlocked()
+    {
+        // If the wallet's signature verfication passes, the wallet must be unlocked.
+        require(
+            msg.sender == address(this) ||
```

Comment:
Please see my comments in `executeMetaTx` inside "MetatxLib". 
It seems even if `msg.sender == address(this)`,we should also check the wallet is not locked.

Reply:
I think both approaches for checking for locked wallets is quite similar, but a small benefit of checking if the wallet is locked or not on the meta tx level is that using wallet funds to pay for transactions is disallowed. I think this is a benefit for wallet users compared to just checking it on the internal call level where the cal may fail quickly but transactions costs can still be reimbursed.